File: Python/Estudo2/ex35.py
import matplotlib.pyplot as plt
import numpy as np
from scipy.special import spence

from metropolis import metropolis_Fermioes_3D

# desprezei as cenas de EF2 

if __name__ == '__main__':

    nmax = 30 # 60
    N = 100 # 200
    nmedidas = 2000 # 18000
    nequi = 2000 # 2000

    Tf = ((3*N)/(4*np.pi))**(2/3)
    Ef = ((3*N)/(4*np.pi))**(2/3)

    # Ts = np.arange(Tf/20 , 2*Tf, Tf/20)
    Ts = np.arange(2*Tf/5 , 2*Tf, Tf/5)
    nTs = len(Ts)

    Emeds = np.zeros(nTs)
    E2meds = np.zeros(nTs)
    nkmed = np.zeros((nmax**3,nTs))
    Ef2s = np.zeros(nTs)
    mu = np.zeros(nTs)

    for i,T in enumerate(Ts):
        print("Simulação",i+1,"T",T)
        Emeds[i], E2meds[i], nkmed[:,i], Ef2s[i] = metropolis_Fermioes_3D(T,nequi,nmedidas,N,nmax)
        mu[i]=3/4+T*np.log(nkmed[0,i]/(1-nkmed[0,i]))
        print(f'T={T}, <E>={Emeds[i]-3/4},  mu={mu[i]}')
    # mu=3/4+Ts*np.log(nkmed[0,:]/(1-nkmed[0,:]))
    print(mu)

    Ef2=Ef2s[-1] # perceber pq?

    Cv = (E2meds - Emeds**2) / (Ts**2)

    # Temperaturas
    Tt = np.linspace(Tf/100 , 2*Tf, 100)
    Tt1 = Tt[:25]
    Tt2 = Tt[75:]
    Tf2 = Ef2

    # Valores teoricos para T<Tf
    Et = N*3*Ef/5*(1+5*np.pi**2/12*(Tt1/Tf)**(5/2))
    mut = Ef*(1-np.pi**2/12*(Tt1/Tf)**2)
    Cvt = N*(np.pi**2/2)*(Tt1/Tf)

    # Valores teoricos para Gas ideal classico
    EGI=3*N*Tt2/2
    muGI=Tt2*np.log(4/(np.sqrt(np.pi)*3)*(Tf/Tt2)**(3/2))
    CvGI=3*N*np.ones(len(Tt2))/2

    plt.figure(1)
    plt.plot(Ts/(Tf2-3/4),(Emeds-3*N/4)/(N*Ef2),'kx')
    plt.plot(Tt1/Tf,Et/(N*Ef),'r-',Tt2/Tf,EGI/(N*Tf),'g-')

    plt.figure(2)
    plt.plot(Ts/(Tf2-3/4),mu/(Ef2-3/4),'kx')
    plt.plot(Tt1/Tf,mut/Ef,'r-', Tt2/Tf,muGI/Ef,'g-')

    plt.figure(3)
    plt.plot(Ts/(Tf2-3/4),Cv/N,'k.',Tt1/Tf,Cvt/N, 'r-')
    plt.plot(Tt2/Tf,CvGI/N,'g-')

    plt.show()

File: Matlab/ExercíciosCarlotaTeste2/Ex35/Ex35.m
close all; clear all;
% Exercicio 35

N = 200;
npassos=20000;
nequi=2000;
nmedidas=npassos-nequi;
nmax=60;
TF=(3*N/(4*pi))^(2/3); % temperatura de fermi
% Energia de Fermi - exp teorica
EF=(3*N/(4*pi))^(2/3);
Tmax=2*TF;
Tv=[TF/20:TF/20:Tmax];
ic=0;

for T=Tv
    ic=ic+1;
    fprintf(1, 'Simulacao %d\n',ic)
    [Emedio(ic),E2medio(ic),nkmedio,EF2]=metropolis_ex35(T,nequi,nmedidas,N,nmax);
    mu(ic)=3/4+T*log(nkmedio(1)/(1-nkmedio(1))); 
    fprintf(1,'T=%f, <E>=%f,  mu=%f\n', T, Emedio(ic)-3/4,  mu(ic));
end

% Capacidade térmica
Cv=(E2medio-Emedio.^2)./Tv.^2;

% Temperaturas
Tt=transpose(linspace(TF/100,Tmax,100));
Tt1=Tt(Tt<=0.5*TF);
Tt2=Tt(Tt>1.5*TF & Tt<=Tmax);
TF2=EF2;

% Valores teoricos (T<TF)
Et=N*3*EF/5*(1+5*pi^2/12*(Tt1/TF).^(5/2));
mut=EF*(1-pi^2/12*(Tt1/TF).^2);
Cvt=N*(pi^2/2)*(Tt1/TF);

% Gás ideal clássico
EGI=3*N*Tt2/2;
muGI=Tt2.*log(4/(sqrt(pi)*3)*(TF./Tt2).^(3/2));
CvGI=3*N*ones(length(Tt2),1)/2;

% Plots
figure(1)
plot(Tv/(TF2-3/4),(Emedio-3*N/4)/(N*EF2),'kx',Tt1/TF,Et/(N*EF),'r-',Tt2/TF,EGI/(N*TF),'g-')
% subtrai-se 3N/4 porque tdos os estados de energia num sistema sem
% condicoes fronteira periodicas estao deslocados 3/4 de energia
% relativamente a um sistema com condicoes fronteira periodicas e as
% formulas teoricas assumem condicoes fronteira periodicas.
title('Energia')
xlabel('T/TF'); ylabel('<E>/N')
legend('Solução','T<Tc','Gás ideal clássico')

figure(2)
plot(Tv/(TF2-3/4),mu/(EF2-3/4),'kx',Tt1/TF,mut/EF,'r-', Tt2/TF,muGI/EF,'g-')
title('potencial químico')
xlabel('T/TF'); ylabel('mu/EF')
legend('Solução','T<Tc','Gás ideal clássico')

figure(3)
plot(Tv/(TF2-3/4),Cv/N,'k.',Tt1/TF,Cvt/N, 'r-',Tt2/TF,CvGI/N,'g-')
title('Capacidade térmica')
xlabel('T/TF'); ylabel('Cv/N')
legend('Solução','T<Tc','Gás ideal clássico')







